tlatools/src/tlc2/tool/Simulator.java

```diff
@@ -103,41 +114,303 @@ public Simulator(String specFile, String configFile, String traceFile, boolean d
 	private final Action[] impliedActions; // the implied-actions to be checked
 	private final boolean checkDeadlock; // check deadlock?
 	private final boolean checkLiveness; // check liveness?
-	private long numOfGenStates;
+	
+	// The total number of states generated by all workers. Written to concurrently.
+	private AtomicLong numOfGenStates;
+	
 	// private Action[] actionTrace; // SZ: never read locally
 	private final String traceFile;
+
+	// The maximum length of a simulated trace.
 	private final long traceDepth;
+
+	// The maximum number of total traces to generate.
 	private final long traceNum;
+
+	// The number of worker threads to use for simulation.
+	private int numWorkers = 1;
+
 	private final RandomGenerator rng;
 	private final long seed;
 	private long aril;
 	private final ObjLongTable astCounts;
 	private boolean isCancelled; // SZ Feb 24, 2009: cancellation added
 	private Value[] localValues = new Value[4];
 
+	// The set of all initial states for the given spec. This should be only be
+	// computed once and re-used whenever a new random trace is generated. This
+	// variable should not be written to concurrently, but is allowed to be read
+	// concurrently.
```

Comment:
It might makes sense to support TLC's continue option to keep state exploration going even if an error is found.